src/pool/pool_jemalloc.c

```diff
@@ -0,0 +1,345 @@
+
+    return false;
+}
+
+static extent_hooks_t arena_extent_hooks = {
+    .alloc = arena_extent_alloc,
+    .dalloc = arena_extent_dalloc,
+    .destroy = arena_extent_destroy,
+    .commit = arena_extent_commit,
+    .decommit = arena_extent_decommit,
+    .purge_lazy = arena_extent_purge_lazy,
+    .purge_forced = arena_extent_purge_forced,
+    .split = arena_extent_split,
+    .merge = arena_extent_merge};
+
+static void *je_malloc(void *pool, size_t size) {
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    unsigned arena = je_pool->arena_zero;
+    int flags = MALLOCX_ARENA(arena) | MALLOCX_TCACHE_NONE;
+    void *ptr = mallocx(size, flags);
+    if (ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return ptr;
+}
+
+static enum umf_result_t je_free(void *pool, void *ptr) {
+    (void)pool; // unused
+
+    if (ptr != NULL) {
+        dallocx(ptr, MALLOCX_TCACHE_NONE);
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void *je_calloc(void *pool, size_t num, size_t size) {
+    size_t csize = num * size;
+    void *ptr = je_malloc(pool, csize);
+    if (ptr == NULL) {
+        // TLS_last_allocation_error is set by je_malloc()
+        return NULL;
+    }
+
+    memset(ptr, 0, csize);
+    return ptr;
+}
+
+static void *je_realloc(void *pool, void *ptr, size_t size) {
+    if (size == 0 && ptr != NULL) {
+        dallocx(ptr, MALLOCX_TCACHE_NONE);
+        TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+        return NULL;
+    } else if (ptr == NULL) {
+        return je_malloc(pool, size);
+    }
+
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    unsigned arena = je_pool->arena_zero;
+    void *new_ptr =
+        rallocx(ptr, size, MALLOCX_ARENA(arena) | MALLOCX_TCACHE_NONE);
+    if (new_ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return new_ptr;
+}
+
+static void *je_aligned_alloc(void *pool, size_t size, size_t alignment) {
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    unsigned arena = je_pool->arena_zero;
+    int flags =
+        MALLOCX_ALIGN(alignment) | MALLOCX_ARENA(arena) | MALLOCX_TCACHE_NONE;
+    void *ptr = mallocx(size, flags);
+    if (ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return ptr;
+}
+
+static enum umf_result_t je_initialize(umf_memory_provider_handle_t hProvider,
+                                       void *params, void **out_pool) {
+    (void)params; // unused
+
+    extent_hooks_t *pHooks = &arena_extent_hooks;
+    size_t unsigned_size = sizeof(unsigned);
+    int err;
+
+    jemalloc_memory_pool_t *pool = malloc(sizeof(jemalloc_memory_pool_t));
+    if (!pool) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    pool->arena_map_len = 1;
```

Comment:
Please either add a TODO here or remove the loop below

Reply:
Changed